### PR TITLE
[IMP] read amounts in user's currency directly

### PR DESCRIPTION
**Pull request by hbrunn**
Assuming most invoices are in the user's currency, building a partner form is orders of magnitude faster this way

### CODE REVIEW

File: addons/account/partner.py [hunk 1]
```diff
@@ -241,11 +241,28 @@ def _debit_search(self, cr, uid, obj, name, args, context=None):
...
+            all_partner_ids = self.pool['res.partner'].search(
+                cr, uid, [('id', 'child_of', partner_id)], context=context)
+            cr.execute(
+                'select sum(price_total) '
+                'from account_invoice_report '
+                'where '
+                'partner_id in %s and state not in %s and currency_id=%s',
+                (tuple(all_partner_ids), tuple(['draft', 'cancel']),
+                 user.company_id.currency_id.id))
+            result[partner_id] = sum(s or 0.0 for s, in cr.fetchall()) or 0.0
```
Comment (mart-e) [inline addons/account/partner.py]: You can remove the or `or 0.0` as `sum` gives zero for an empty list.

**Comment by mart-e** [comment]
Hello,

This seems kind of hacky. Is it the method for the computation of `user_currency_price_total` that is slow ? (so probably fix this one).
Is the improvement because you are skipping computation for the invoices in the same currency or because you are skipping the orm ?

**Comment by hbrunn** [comment]
@mart-e my profiling scripts says they are both slow (this is from the original):
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   2086/4    0.003    0.000   10.547    2.637 ./openerp/api.py:236(wrapper)
     41/2    0.001    0.000   10.547    5.274 ./openerp/models.py:3137(read)
    41/10    0.003    0.000   10.543    1.054 ./openerp/models.py:3236(_read_from_database)
    20/18    0.000    0.000   10.522    0.585 ./openerp/osv/fields.py:1445(get)
      323    0.000    0.000   10.473    0.032 ./openerp/sql_db.py:151(wrapper)
      182    0.001    0.000   10.473    0.058 ./openerp/sql_db.py:220(execute)
      182   10.471    0.058   10.472    0.058 {method 'execute' of 'psycopg2.extensions.cursor' objects}
        2    0.000    0.000   10.402    5.201 /home/hbrunn/git-workdirs/ocb-8.0/addons/account/partner.py:241(_invoice_total)
  826/398    0.001    0.000   10.376    0.026 ./openerp/fields.py:797(__get__)
    31/21    0.000    0.000   10.373    0.494 ./openerp/fields.py:884(determine_value)
    29/23    0.001    0.000   10.373    0.451 ./openerp/models.py:3186(_prefetch_field)
       22    0.000    0.000   10.348    0.470 {sum}
        4    0.000    0.000   10.348    2.587 /home/hbrunn/git-workdirs/ocb-8.0/addons/account/partner.py:248(<genexpr>)
        2    0.000    0.000    5.113    2.556 /home/hbrunn/git-workdirs/ocb-8.0/addons/account/report/account_invoice_report.py:32(_compute_amounts_in_user_currency)
```
So yes, with fixing `_compute_amounts_in_user_currency` (this and `res.currency#compute` probably should use the v8 api to have better caching) we also come far, but we then still have a lot of overhead.

I can provide a bigger patch that makes more use of caching if you want, that should solve the problem too.

**Comment by hbrunn** [comment]
I use this to find out where to look for performance tweaks: https://gist.github.com/hbrunn/c7f7fbab47261475772c

**Comment by mart-e** [comment]
After more analysis, it's actually the SQL view `account_invoice_report` that is slow like hell.
Check made on our database with thousands of invoices:
```
# select sum(price_total) from account_invoice_report where partner_id in (197805, 387420);
Time: 23.014 ms
# select sum(price_total) from account_invoice_report where id in (39745, 43645);
Time: 883.365 ms
```
When you are using the ORM, the generated queries are using the `id in` query which makes slow join.

Checking what could be the best solution.

File: addons/account/partner.py (hunk 1 above)
Comment (mart-e) [inline addons/account/partner.py]: you will get at most only one record so no need for a `sum`, you can use `fetchone`

File: addons/account/partner.py [hunk 2]
```diff
@@ -241,11 +241,28 @@ def _debit_search(self, cr, uid, obj, name, args, context=None):
...
+            # and then add invoices in other currencies
+            domain = [
+                ('partner_id', 'in', all_partner_ids),
+                ('state', 'not in', ['draft', 'cancel']),
+                ('currency_id', '!=', user.company_id.currency_id.id),
+            ]
             invoice_ids = account_invoice_report.search(cr, uid, domain, context=context)
             invoices = account_invoice_report.browse(cr, uid, invoice_ids, context=context)
```
Comment (mart-e) [inline addons/account/partner.py]: probably worth a `if invoice_ids:` as you comment stated

File: addons/account/partner.py (hunk 1 above)
Comment (gdgellatly) [inline addons/account/partner.py]: I am concerned this query does not handle multi company correctly, a
scenario which could return more than one record, but shouldn't.

In addons/account/partner.py
<https://github.com/odoo/odoo/pull/6654#discussion_r30240331>:

> @@ -241,11 +241,28 @@ def _debit_search(self, cr, uid, obj, name, args, context=None):
>      def _invoice_total(self, cr, uid, ids, field_name, arg, context=None):
>          result = {}
>          account_invoice_report = self.pool.get('account.invoice.report')
> -        for partner in self.browse(cr, uid, ids, context=context):
> -            domain = [('partner_id', 'child_of', partner.id), ('state', 'not in', ['draft', 'cancel'])]
> +        user = self.pool['res.users'].browse(cr, uid, uid, context=context)
> +        for partner_id in ids:
> +            # first read invoices in the user's currency
> +            all_partner_ids = self.pool['res.partner'].search(
> +                cr, uid, [('id', 'child_of', partner_id)], context=context)
> +            cr.execute(
> +                'select sum(price_total) '
> +                'from account_invoice_report '
> +                'where '
> +                'partner_id in %s and state not in %s and currency_id=%s',
> +                (tuple(all_partner_ids), tuple(['draft', 'cancel']),
> +                 user.company_id.currency_id.id))
> +            result[partner_id] = sum(s or 0.0 for s, in cr.fetchall()) or 0.0

you will get at most only one record so no need for a sum, you can use
fetchone

—
Reply to this email directly or view it on GitHub
<https://github.com/odoo/odoo/pull/6654/files#r30240331>.

**Comment by hbrunn** [comment]
@mart-e I fixed your comments.
As for a more elegant solution using the ORM: I currently can't think of something

**Comment by mart-e** [comment]
I just noticed another potential issue with security rules. You do not respect multicompany rules (and potential other security rules) in the SQL query.
Check what we do [in res.partner name_search](https://github.com/odoo/odoo/blob/7360a30b27befb1009ff58372bd0a47f8c2209e1/openerp/addons/base/res/res_partner.py#L664) how we build a query respecting security rules.
Except that, looks good to me.

**Comment by mart-e** [comment]
@hbrunn I have created another PR at #7053 with a global query that should work also for invoices in a different currency. This should make it faster for every partner.

**Comment by TheCloneMaster** [comment]
@hbrunn are you proposing this for OCB?

**Comment by hbrunn** [comment]
@TheCloneMaster it's in fa3ace7, which has been synced to OCB a while ago already

**Pull request closed without merging**